keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,575 @@
+
+
+### Add .status.nextScheduleTime
+[#issue/78564](https://github.com/kubernetes/kubernetes/issues/78564)
+Add `nextScheduleTime` to `.status` that tracks the next time the job will be scheduled. This may not be accurate with `Forbid` concurrency policy. This only tracks the `Job` creation time and not the actual `Pod` creation time.
+
+| Concurrency policy | notes |
+|:----------|:----------|
+| Allow   | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter |
+| Forbid  |  `nextScheduleTime` would not be accurate if the previous `Job` takes longer than the cron interval time |
+| Replace | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter along with older concurrent `Job` cleanup if applicable. |
+
+
+
+### Add Counters
+These counters would be added to `.status` section of the CronJob object:
+- `SuccessfulRuns` Count of all successful runs
+- `FailedlRuns` Count of all failed runs
+- `FailuresSinceSuccess` Count of all successful runs since last failure
+
+
+
+### Add .status.conditions
+Add a condition array with `Settled` condition type. This would help with the effort of standarizng conditions across all core types. `Settled` is set at the end of every successful reconcile run. The key thing to note here is the notions of `Settled` does not imply the `Job`s are running correctly. It just means that the controller is done processing this object successfully.
+
+
+### Support Jitter for cronjobs
+We propose to introduce `.spec.jitter` which is a percentage of the time delta to the next schedule. We propose to cap it to 50%. There is also a [community request](https://github.com/kubernetes/community/issues/2440) for this.
+
+```golang
+delta = nextScheduleTime - currentTime
+jitter = delta*cronjob.Spec.Jitter/100
+nextScheduleTime += jitter
+```
+
+### Fix applicable open issues
+These are the [current](https://github.com/kubernetes/kubernetes/issues/82659) list of issues that are being targeted for GA. 
+
+- [Updating a cronjob causes jobs to be scheduled retroactively](https://github.com/kubernetes/kubernetes/issues/63371)
+- [CLI: Updated CronJob Schedule Missing from Dry Run](https://github.com/kubernetes/kubernetes/issues/73613)
+- [Kubernetes CronJob pods is not getting clean-up when Job is completed](https://github.com/kubernetes/kubernetes/issues/74741)
+- [Infinite ImagePullBackOff CronJob results in resource leak](https://github.com/kubernetes/kubernetes/issues/76570)
+- [Cronjob `spec.schedule` cannot be change when `spec.schedule` value not `"` or `’`](https://github.com/kubernetes/kubernetes/issues/78646)
+- [Kubelet CPU/Memory Usage linearly increases using CronJob](https://github.com/kubernetes/kubernetes/issues/64137)
+- [Stopping cluster overnight prevents scheduled jobs from running after cluster startup](https://github.com/kubernetes/kubernetes/issues/42649)
```

Comment:
https://github.com/kubernetes/kubernetes/pull/81557 fixes this and has been pending review for over half a year by now.

Reply:
Thanks for bringing it to our notice. Will review agains.